### PR TITLE
move apply from apply to aggregate

### PR DESCRIPTION
**Pull request by thewilsonator**
apply is only used by aggregate.d and dclass.d and doesn't really belong in apply which should probably be named postorderVisitor

### CODE REVIEW

**Comment by thewilsonator** [comment]
urgh
```
src\dmd\tocvdebug.d(561): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(561):        cannot pass argument `& cv_mem_count` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(572): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(572):        cannot pass argument `& cv_mem_p` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(724): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(724):        cannot pass argument `& cv_mem_count` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(775): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(775):        cannot pass argument `& cv_mem_p` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
```

File: src/dmd/aggregate.d [hunk 1]
```diff
@@ -866,4 +863,38 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
...
+}
+
+/*********************************
+ * Iterate this dsymbol or members of this scoped dsymbol, then
+ * call `fp` with the found symbol and `params`.
+ * Params:
+ *  symbol = the dsymbol or parent of members to call fp on
+ *  fp = function pointer to process the iterated symbol.
+ *       If it returns nonzero, the iteration will be aborted.
+ *  params = any parameters passed to fp.
```
Comment (WalterBright) [inline src/dmd/aggregate.d:876]: it's been changed to `agg`

Comment (thewilsonator) [reply src/dmd/aggregate.d:876]: I'll be changing that back soon

**Review by RazvanN7 — APPROVED** [review]

**Comment by thewilsonator** [comment]
@RazvanN7 pong. Apologies for the late response.

File: src/dmd/aggregate.d [hunk 2]
```diff
@@ -814,3 +811,36 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
...
+ * Params:
+ *  symbol = the dsymbol or parent of members to call fp on
+ *  fp = function pointer to process the iterated symbol.
+ *       If it returns nonzero, the iteration will be aborted.
+ *  ctx = context parameter passed to fp.
+ * Returns:
+ *  nonzero if the iteration is aborted by the return value of fp,
+ *  or 0 if it's completed.
+ */
+int apply(Dsymbol symbol, int function(Dsymbol, void*) fp, void* ctx)
```
Comment (RazvanN7) [inline src/dmd/aggregate.d:827]: Why not use the variadic parameter? It looks nicer than having `void*` in the signature. Are you worried about the size of the binary? It seems like there are 3 different instantiations only.

Comment (RazvanN7) [reply src/dmd/aggregate.d:827]: @thewilsonator  ?

Comment (RazvanN7) [reply src/dmd/aggregate.d:827]: Still, I think it's better to use the variadic. But I am willing to let this slide.

**Comment by thewilsonator** [comment]
Thanks!

**Pull request merged**